File: Leetcode/AddBinary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        string res;
        int la = a.length() - 1, lb = b.length() - 1;
        int up = 0, I = 0;
        while(la >= 0 && lb >= 0){
            int g = a[la--] - '0' + b[lb--] - '0' + up;
            up = g / 2;
            g %= 2;
            res.push_back(g + '0');
        }
        while(la >= 0){
            int g = a[la--] - '0' + up;
            up = g / 2;
            g %= 2;
            res.push_back(g + '0');
        }
        while(lb >= 0){
            int g = b[lb--] - '0' + up;
            up = g / 2;
            g %= 2;
            res.push_back(g + '0');
        }
        
        if(up) res.push_back('1');
        reverse(res.begin(), res.end());
        return res;
    }
};

File: Leetcode/AddBinary.rs
// https://leetcode.com/problems/add-binary

struct Solution;

impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        use std::iter::repeat;

        let n = a.len().max(b.len());

        let (carry, mut sum) = a
            .into_bytes()
            .into_iter()
            .rev()
            .chain(repeat(b'0'))
            .zip(b.into_bytes().into_iter().rev().chain(repeat(b'0')))
            .take(n)
            .fold((0, Vec::new()), |(mut carry, mut sum), (x, y)| {
                let x = x - b'0';
                let y = y - b'0';

                sum.push((x ^ y ^ carry) + b'0');
                carry = x & y | (carry & (x ^ y));

                (carry, sum)
            });

        if carry == 1 {
            sum.push(b'1');
        }
        sum.reverse();

        String::from_utf8(sum).unwrap()
    }
}
